### ISSUE TITLE
ENH: Kullback-Leibler divergence and crossentropy for continious distributions and their derivatives and hessians

### ISSUE THREAD

**Issue by KOLANICH** [scipy/scipy#15974]
### Is your feature request related to a problem? Please describe.

Let we have an approximation to PDF provided pointwise. We want to fit distribution to that PDF. One of the ways is to minimize KL-divergence or equivalently, crossentropy.

### Describe the solution you'd like.

`rv_continious().crossentropy(x)`
`rv_continious().KL_divergence(x)`

where `x` can be either an array representing PDF, or a frozen distribution. In the case of distributions formulas in closed form should be used.

There also gradients and hessians should be available, so we could avoid estimating them numerically.

### Describe alternatives you've considered.

Having methods to fit the PDF for each distribution. It is surely needed, but it is a bit different thing. This feature is likely a prereq for that.

**Comment by tupui** [comment]
Hi @KOLANICH, thank you for reporting. I am not sure I understand what you are proposing. We have some statistical distances and to me your proposal reads as adding 2 more. Correct?

**Comment by mdhaber** [comment]
The suggestion is to add them as methods of distributions, right? This would be different from the functions, which act on two samples. Here, one (or both) of the samples would be replaced with a distribution.

**Comment by tupui** [comment]
Oh if it's really about adding this to the interface of `rv_continuous` then I am -1.

**Comment by tupui** [comment]
For the continuous vs discrete computation, is this really useful in practice?

**Comment by mdhaber** [comment]
I don't actually _use_ SciPy much, but actually this is something I would have used in a study before. I wanted some notion of KL divergence between a distribution and a sample.

**Comment by tupui** [comment]
When I had to do this, I was just using a KDE to work things out. See `moment_independent` in https://github.com/tupui/batman/blob/master/batman/visualization/density.py (old code be kind 😉)

**Comment by rkern** [comment]
Yeah, when people compute KL divergence between a notional parametric distribution and "some data", there is a middle step of density estimation, then you can numerically integrate the density estimate with the parametric distribution. You can do that density estimate in any number of ways, but some will be more convenient for the numerical integration in addition to the set of choices one needs to make to model the data faithfully. Sometimes KDEs work nicely (particularly when you can take shortcuts with the integration), other times histograms (a different set of shortcuts), other times something else. So yeah, it's a thing that people do, but with so many variations that I don't think it's a good fit for the `rv_continuous` interface.

For people who are interested in doing variational inference like this, it is indeed useful to have the derivatives so you can do optimization. But you probably also want more sophisticated models built in a more complete probablistic programming language like PyMC or Stan using their underlying autodiff capabilities to do it. I think it's a lot of effort to add all this to the `rv_continuous` framework for not enough benefit because the "models" that we could then fit (i.e. just the plain parametric distributions that we have) are not very expressive.

I think we're reaching a consensus among the scipy maintainers that we're all -1 on this.

**Comment by mdhaber** [comment]
> I think we're reaching a consensus among the scipy maintainers that we're all -1 on this.

Are we -1 on all of this issue, or just parts of it @rkern @tupui? I see the strongest arguments against only part of it.

---

The desire to have partial derivatives information for distributions is tracked in gh-15928, so let's restrict this issue to the ideas about relative entropy (/KL-divergence) and cross entropy.

Let's also separate the two explicit suggestions at the top:

1. relative and cross entropy between two continuous distributions

2. relative and cross entropy between continuous distribution and data

and the two natural extensions of these ideas:

3. relative and cross entropy between two discrete distributions

4. relative and cross entropy between discrete distribution and data

into four separate ideas and consider them individually.

1. Continuous-Continuous case: seems to me like this could be in scope, actually. The entropy of many distributions is calculated numerically, in which case `entropy` is really convenience method so the user doesn't have to call `quad` directly. Do we regret offering the `entropy` method? I don't see why methods for computing relative and cross entropy numerically would be so different.

2. Continuous-Data case: this seems to have the most arguments against it. Yes, I'd agree that I'm -1 on special support for this. Instead, I'd say users needs to turn their data into a distribution or `gaussian_kde`, then treat it like case 1.

3. Discrete-Discrete case: this is like case 1. For two distributions with compact support, computing the quantities from their definitions seems like it would work well enough. For distributions with infinite support, a method for approximating the calculations would be valuable. (Do we have a function like [nsum](https://mpmath.org/doc/current/calculus/sums_limits.html)? That would be very valuable in its own right.)

4. Discrete-Data case: probably more reasonable than 2 to have special support for this, since turning data into a discrete distribution seems more straightforward than converting data into a continuous distribution. Still, it is probably better to just require the user to create their own discrete distribution (e.g. `rv_histogram`) and treat this like case 3. We could show an example in the docs.

**Comment by rkern** [comment]
I'm -1 on all of it, at least as they pertain to adding new extensions to the `rv_continuous/rv_discrete` interfaces. The distribution-distribution cases are reasonable to implement as standalone functions, maybe, if anybody actually needs it. But the core of the request was about the data cases, as evidenced by what the requester actually implemented in their own code, so that's not a big motivation to work on the distribution-distribution cases, IMO.

**Comment by tupui** [comment]
I was just -1 on expanding the `rv_*` interfaces. Having convenience methods that do numerical integration to get entropy has some value.

**Comment by mdhaber** [comment]
OK let's keep this in mind in gh-15928, but the conclusion here seems to be that we do not want to expand the existing distribution infrastructure with these methods, and we do not envision that these methods would accept data `x` - only distribution-like objects with `pdf`/`logpdf` methods (albeit possibly generated from data).
